tests/concurrent_client_request_test.go

```diff
@@ -52,39 +55,103 @@ func getTestClient(front string, t *testing.T) *http.Client {
 	}
 }
 
+type backend struct {
+	mu      sync.Mutex // mu protects conn
+	conn    agent.AgentService_ConnectServer
+	agentID string
+}
+
+func (b *backend) Send(p *clientproto.Packet) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	return b.conn.Send(p)
+}
+
+func (b *backend) Context() context.Context {
+	// TODO: does Context require lock protection?
+	return b.conn.Context()
+}
+
+func (b *backend) AgentID() string {
+	return b.agentID
+}
+
+func newBackend(agentID string, conn agent.AgentService_ConnectServer) *backend {
+	return &backend{agentID: agentID, conn: conn}
+}
+
 // singleTimeManager makes sure that a backend only serves one request.
 type singleTimeManager struct {
 	mu       sync.Mutex
-	backends map[string]agent.AgentService_ConnectServer
+	backends map[string][]*backend
```

Comment:
How about `map[string][]Backend`? Then you don't need to duplicate the `backend` code.

You might find it difficult to implement the `RemoveBackend` method if you use `Backend`. If so, we can skip the implementation of `RemoveBackend`, which is not used in this test anyway.

Reply:
The newBackend function is private, so we'll still need to duplicate the `backend` code I think.

Reply:
Can we make newBackend function to public? That sounds a common pattern, a public function that returns a public interface implemented by a private object.